### PR TITLE
upgrade to Rocket.Chat 0.46.0 (master)

### LINKED ISSUE #65: sync upstream changes
should make things run way faster in development

### PR DESCRIPTION
**Pull request by jeffreywescott**
## PR

Fixes #65.

### **Important:**  About This Branch

**RELEVANT commits begin below on Nov 28, 2016.**

**TL;DR**: This is an entirely new branch in our fork of the repository. We will rename it to `lg-master` rather than trying to merge it. A merge attempt will fail. We'll save the _old_ `lg-master` branch with a date-stamp appended to it in case we need to look at commit history for a date earlier than when the rename occurs.

**Background:** The reason is that, when I first set us up on Rocket.Chat, I did so from their `develop` branch. I've since discovered that (duh) that branch is considered "unstable". However, we want to base our changes on their `master` branch. Unfortunately, commit history upstream between those two branches is not compatible (presumably from some `git push --force` in the `develop` branch), so I was not able to find a shared ancestor commit.

Fortunately, almost all of our changes have been isolated in our own packages, the notable exceptions being a few things in the root directory (e.g., `package.json`, `.codeclimate.yml`, etc.), and changes to the `packages` and `versions` files in the `.meteor` subdirectory. So building this branch was fairly straight forward.

#### To-Do After Branch Renames

- [x] change Codeship commands for testing (npm script names prefixed with `lg:`)
- update CodeClimate
  - [x] change default branch
  - [x] CodeClimate was added by RC upstream, also, so update CodeClimate settings so that we don't get "dinged" for code that's not ours

#### Deploying

We will need to:

- [x] create a new `lg-echo` deployment (plus DB and user) on compose.io (using the "WiredTiger" storage engine)
- [x] update `MONGO_URL`
- [x] set buildpack back to [standard Rocket.Chat buildpack](https://github.com/RocketChat/heroku-buildpack-meteor.git)
- [x] restore from backup
- [x] delete old `lg-chat` deployment on compose.io
- [x] delete `echo-chat-staging` deployment on compose.io
- [x] spin-down dynos for `lg-echo-staging` on Heroku

This means, of course, that we'll have to update the Mongo environment variable with the new DB info.

And, to avoid the 60-second "boot timeout" on Heroku, we will need to run the migrations first (by starting the server):

- [x] create a `Procfile` with a "bogus" `web:` process type so that the server won't start, then:

```
local$ heroku run bash
...
remote$ .meteor/heroku_build/bin/node .meteor/heroku_build/app/main.js
```

- [x] remove the `Procfile`

##### Staging

Because it is such a significant upgrade, I created a staging environment so that I could preview everything and ensure the storage engine changes (see Deploying above), database migrations, and other things would work. You can check things out, too, if you want: https://lg-echo-staging.herokuapp.com/

##### Change File Limits

The new version needs more open files than are permitted by default on OS X. You'll probably need to do this from your terminal:

```
echo kern.maxfiles=65536 | sudo tee -a /etc/sysctl.conf
echo kern.maxfilesperproc=65536 | sudo tee -a /etc/sysctl.conf
sudo sysctl -w kern.maxfiles=65536
sudo sysctl -w kern.maxfilesperproc=65536
ulimit -n 65536 65536    
```

And, you should add this to your shell initialization script (e.g., `~/.bash_profile`):

```
ulimit -n 65536 65536
```

##### Miscellaneous Stuff

You'll want to do this locally before you run the first time:

```
meteor npm install --save bcrypt
```

It's also worth noting that the scripts in `package.json` were prefixed with `lg:` to avoid upstream collisions. Previous versions of Rocket.Chat didn't use a `package.json` file, but since they're on Meteor 1.4 now, we're in a more "node-like" environment. So, you'll need to use `npm run lg:start` to launch the server.

#### Troubleshooting

- If you get weird errors running locally, try removing your previous cached Meteor installation, then starting the server again: `rm -rf ~/.meteor ; npm run lg:start`

### CODE REVIEW

**Comment by bundacia** [comment]
Looks like maybe we need to update our README to reflect that we're branching off upstream master not, not develop?

**Comment by bundacia** [comment]
locally rocketchat is sunning on port 25610 now, instead of 3000. Do we need to update the README?

**Review by bundacia — APPROVED** [review]
Aside from the minor comments I made this looks FANTASTIC! So excited to get on the new version. Works great locally and (so far) is not pegging my CPU the way it used to.

LGTM

**Comment by jeffreywescott** [comment]
Interesting about the port. Not sure what's going on there ... I'll try to get it to be a consistent port at least; it seems to be changing between server restarts right now.

**Comment by jeffreywescott** [comment]
On second thought ... it _does_ seem to be running on port 3000 after all, despite the `Process Port: 28128` message. Strange.

**Comment by jeffreywescott** [comment]
Also, the stuff in the readme about the `develop` branch is still correct -- it relates to sending PRs upstream, which still has to happen from that branch.

**Comment by bundacia** [comment]
I've confirmed that this is not running on port 3000 for me locally. I had to update my mehserve config.

**Comment by bundacia** [comment]
oh, actually, maybe it's running on both? Proxying 3000 back to the other port (which seems to be different every time)

**Comment by jeffreywescott** [comment]
So, my mehserve is configured to 3000, and it works, even when the port says 28128 in the output message. Are you still seeing it _not_ working?

**Comment by bundacia** [comment]
no, sorry I wasn't clear. It IS working with mehserve set to port 3000 now. It seems to be using 2 ports, but one of them is 3000. All good.
